src/main/scala/com/fulcrumgenomics/bam/CallConsensusFromUmis.scala

```diff
@@ -0,0 +1,450 @@
+                     errorRatePreUmi: PhredValue             = DefaultErrorRatePreUmi,
+                     errorRatePostUmi: PhredValue            = DefaultErrorRatePostUmi,
+                     maxBaseQuality: PhredValue              = DefaultMaxBaseQuality,
+                     baseQualityShift: PhredValue            = DefaultBaseQualityShift,
+                     minConsensusBaseQuality: PhredValue     = DefaultMinConsensusBaseQuality,
+                     minReads: Int                           = DefaultMinReads,
+                     minMeanConsensusBaseQuality: PhredValue = DefaultMinMeanConsensusBaseQuality
+                    ): Option[(String, String)] = {
+    // TODO: if the reads are mapped to the negative strand, reverse complement so they are in sequencing order
+    val bases = records.map { rec => rec.getReadString.toUpperCase }
+    val quals = records.map { rec => rec.getBaseQualityString }
+
+    fromBasesAndQualities(
+      basesAndQualities           = bases.zip(quals),
+      errorRatePreUmi             = errorRatePreUmi,
+      errorRatePostUmi            = errorRatePostUmi,
+      maxBaseQuality              = maxBaseQuality,
+      baseQualityShift            = baseQualityShift,
+      minConsensusBaseQuality     = minConsensusBaseQuality,
+      minReads                    = minReads,
+      minMeanConsensusBaseQuality = minMeanConsensusBaseQuality
+    )
+  }
+
+  private[bam] def phredToFastq(calls: Seq[PhredValue]): String = {
+    SAMUtils.phredToFastq(calls.map(pv => Math.min(SAMUtils.MAX_PHRED_SCORE, pv.toInt)).map(_.toByte).toArray[Byte])
+  }
+
+  /** Get the most likely consensus bases and qualities. */
+  private[bam] def consensusCalls(bases: Seq[String],
+                                  quals: Seq[Seq[PhredValue]],
+                                  minConsensusBaseQuality: PhredValue = DefaultMinConsensusBaseQuality,
+                                  minReads: Int = DefaultMinReads): Seq[(Byte, PhredValue)] = {
+    val maxBaseLength = bases.map(_.length).max
+
+    // go through each position in the read(s).
+    Seq.range(0, maxBaseLength, 1).map {
+      baseIdx =>
+        val basesAtIdx = bases.filter { str => baseIdx < str.length}.map { str => str.charAt(baseIdx) }
+
+        // check to see if we heave a enough reads
+        if (basesAtIdx.length < minReads) (SequenceUtil.N, ZeroPhred)
+        else {
+          val qualsAtIdx         = quals.filter { q => baseIdx < q.length}.map { q => q(baseIdx) }
+          val basesAndQualsAtIdx = basesAtIdx.zip(qualsAtIdx)
+
+          // TODO: handle Ns in the reads
+          // get the phred-scaled probability of each potential DNA base
+          var phreds = SequenceUtil.VALID_BASES_UPPER.map(_.toChar).map {
+            candidateBase =>
+              basesAndQualsAtIdx.map {
+                case (base, qual) =>
+                  if (base == candidateBase) qual.inv() // probability of not an error
```

Comment:
what if the quality is zero, this could make the probability zero.  There should be some minimum, say `PhredValue.Precision`?